sourcemap/fleet_store.go

```diff
@@ -71,35 +75,45 @@ func newFleetStore(
 	fleetCfg *config.Fleet,
 	cfgs []config.SourceMapMetadata,
 ) (fleetStore, error) {
-	// TODO(stn): Add support for multiple fleet hosts
-	// cf. https://github.com/elastic/apm-server/issues/5514
-	host := fleetCfg.Hosts[0]
-	fleetURLs := make(map[key]string)
+	sourceMapURLs := make(map[key]string)
+	fleetBaseURLs := make([]string, len(fleetCfg.Hosts))
 
 	for _, cfg := range cfgs {
 		k := key{cfg.ServiceName, cfg.ServiceVersion, cfg.BundleFilepath}
-		u, err := common.MakeURL(fleetCfg.Protocol, cfg.SourceMapURL, host, 8220)
+		sourceMapURLs[k] = cfg.SourceMapURL
+	}
+
+	for i, host := range fleetCfg.Hosts {
+		baseURL, err := common.MakeURL(fleetCfg.Protocol, "", host, defaultFleetPort)
 		if err != nil {
 			return fleetStore{}, err
 		}
-		fleetURLs[k] = u
+		fleetBaseURLs[i] = baseURL
 	}
+
 	return fleetStore{
-		apikey:    "ApiKey " + fleetCfg.AccessAPIKey,
-		fleetURLs: fleetURLs,
-		c:         c,
+		apikey:        "ApiKey " + fleetCfg.AccessAPIKey,
+		fleetBaseURLs: fleetBaseURLs,
+		sourceMapURLs: sourceMapURLs,
+		c:             c,
 	}, nil
 }
 
 func (f fleetStore) fetch(ctx context.Context, name, version, path string) (string, error) {
 	k := key{name, version, path}
-	fleetURL, ok := f.fleetURLs[k]
+	sourceMapURL, ok := f.sourceMapURLs[k]
 	if !ok {
 		return "", fmt.Errorf("unable to find sourcemap.url for service.name=%s service.version=%s bundle.path=%s",
 			name, version, path,
 		)
 	}
 
+	rng := rand.New(rand.NewSource(time.Now().UnixNano()))
```

Comment:
Typically you would create a `*rand.Rand` once and reuse it. Did you intentionally choose to create a new one each time? If not, I'd prefer that we create one in `newFleetStore` and store it in the `fleetStore struct. It would need to be protected with a mutex.

Alternatively, we could use the top-level `rand.Intn` function.

Reply:
Another thing: we shouldn't assume that all of the addresses will be accessible to apm-server. In theory there may be a fleet-server in a network that apm-server cannot access.

I think we should probably try each address until we get a success. In that case it might make more sense to use https://pkg.go.dev/math/rand#Perm

Reply:
Do you want to try each address in series, or in parallel? I'm assuming there will be less than 10 fleet servers to contact, and given the relative infrequency of the lookups, I wouldn't expect resource consumption to be an issue.

Reply:
Trying them in parallel sounds fine.

Maybe later we can do something a little more elaborate, e.g. keep track of the fastest/most recently available server, and add a small delay before trying others. But we probably don't need to do that initially.